File: common/src/main/java/com/ssblur/alchimiae/blockentity/BoilerBlockEntity.java
package com.ssblur.alchimiae.blockentity;

import com.ssblur.alchimiae.events.network.client.ParticleNetwork;
import com.ssblur.alchimiae.item.AlchimiaeItems;
import com.ssblur.alchimiae.screen.menu.BoilerMenu;
import dev.architectury.networking.NetworkManager;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.NonNullList;
import net.minecraft.core.component.DataComponents;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.ContainerHelper;
import net.minecraft.world.WorldlyContainer;
import net.minecraft.world.effect.MobEffectInstance;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.entity.player.StackedContents;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.ContainerData;
import net.minecraft.world.inventory.StackedContentsCompatible;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.alchemy.PotionContents;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity;
import net.minecraft.world.level.block.entity.BaseContainerBlockEntity;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.Vec3;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BoilerBlockEntity extends BaseContainerBlockEntity implements WorldlyContainer, StackedContentsCompatible {
  public static final int INGREDIENT_SLOT = 0;
  public static final int FUEL_SLOT = 1;
  public static final int RESULT_SLOT = 2;
  public static final int PROCESS_TIME = 400;

  int litTime, litDuration, processTime;
  public final ContainerData dataAccess;

  NonNullList<ItemStack> inventory;
  public BoilerBlockEntity(BlockPos blockPos, BlockState blockState) {
    super(AlchimiaeBlockEntities.BOILER.get(), blockPos, blockState);
    inventory = NonNullList.withSize(3, ItemStack.EMPTY);

    this.dataAccess = new ContainerData() {
      public int get(int i) {
        switch (i) {
          case 0 -> {
            return litTime;
          }
          case 1 -> {
            return litDuration;
          }
          case 2 -> {
            return processTime;
          }
          default -> {
            return 0;
          }
        }
      }

      public void set(int i, int j) {
        switch (i) {
          case 0 -> litTime = j;
          case 1 -> litDuration = j;
          case 2 -> processTime = j;
        }

      }

      @Override
      public int getCount() {
        return 3;
      }
    };
    }

  @Override
  public int getContainerSize() {
    return 3;
  }

  @Override
  protected Component getDefaultName() {
    return Component.translatable("menu.alchimiae.boiler");
  }

  @Override
  protected NonNullList<ItemStack> getItems() {
    return inventory;
  }

  @Override
  protected void setItems(NonNullList<ItemStack> nonNullList) {
    inventory = nonNullList;
  }

  @Override
  public boolean isEmpty() {
    return inventory.isEmpty();
  }

  @Override
  public ItemStack getItem(int i) {
    return inventory.get(i);
  }

  @Override
  public ItemStack removeItem(int i, int j) {
    setChanged();

    var stack = inventory.get(i);
    if(j >= stack.getCount()) {
      inventory.set(i, ItemStack.EMPTY);
      return stack;
    }
    stack.shrink(j);
    return stack.copyWithCount(j);
  }

  @Override
  public ItemStack removeItemNoUpdate(int i) {
    var stack = inventory.get(i);
    inventory.set(i, ItemStack.EMPTY);
    return stack;
  }

  @Override
  public void setItem(int i, ItemStack itemStack) {
    inventory.set(i, itemStack);
  }

  @Override
  public boolean stillValid(Player player) {
    return player.distanceToSqr(getBlockPos().getCenter()) < 16;
  }

  @Override
  public void clearContent() {
    inventory = NonNullList.withSize(3, ItemStack.EMPTY);
  }

  @Override
  protected AbstractContainerMenu createMenu(int i, Inventory inventory) {
    return new BoilerMenu(i, inventory, this);
  }

  @Override
  public int[] getSlotsForFace(Direction direction) {
    switch (direction) {
      case DOWN -> {
        return new int[] {RESULT_SLOT};
      }
      case UP -> {
        return new int[] {INGREDIENT_SLOT};
      }
      default -> {
        return new int[] {FUEL_SLOT};
      }
    }
  }

  @Override
  public boolean canPlaceItemThroughFace(int i, ItemStack itemStack, @Nullable Direction direction) {
    return direction == null || Arrays.stream(getSlotsForFace(direction)).anyMatch(s -> s == i);
  }

  @Override
  public boolean canTakeItemThroughFace(int i, ItemStack itemStack, Direction direction) {
    return Arrays.stream(getSlotsForFace(direction)).anyMatch(s -> s == i);
  }

  @Override
  public void fillStackedContents(StackedContents stackedContents) {
    for (ItemStack itemStack : inventory)
      stackedContents.accountStack(itemStack);
  }

  @Override
  protected void saveAdditional(CompoundTag compoundTag, HolderLookup.Provider provider) {
    super.saveAdditional(compoundTag, provider);
    ContainerHelper.saveAllItems(compoundTag, inventory, provider);

    compoundTag.putInt("processTime", processTime);
    compoundTag.putInt("litTime", litTime);
    compoundTag.putInt("litDuration", litDuration);
  }

  @Override
  protected void loadAdditional(CompoundTag compoundTag, HolderLookup.Provider provider) {
    super.loadAdditional(compoundTag, provider);
    ContainerHelper.loadAllItems(compoundTag, inventory, provider);

    processTime = compoundTag.getInt("processTime");
    litTime = compoundTag.getInt("litTime");
    litDuration = compoundTag.getInt("litDuration");
  }

  @SuppressWarnings("unused")
  public static void tick(Level level, BlockPos blockPos, BlockState blockState, BlockEntity blockEntity) {
    if (blockEntity instanceof BoilerBlockEntity entity)
      entity.tick();
  }

  public void tick() {
    if(level == null) return;
    if(litTime > 0) {
      if(litTime % 3 == 0) {
        double x = getBlockPos().getX() + (10d / 16d) + level.random.nextDouble() / 10d;
        double y = getBlockPos().getY() + (5d / 16d);
        double z = getBlockPos().getZ() + (5d / 16d) + level.random.nextDouble() / 10d;
        if (level instanceof ServerLevel serverLevel)
          NetworkManager.sendToPlayers(serverLevel.players(), new ParticleNetwork.Payload(new Vec3(x, y, z), ParticleNetwork.TYPE.FLAME));
      }
      litTime--;
    }

    if(inventory.get(INGREDIENT_SLOT).is(AlchimiaeItems.MASH.get())) {
      if(litTime == 0 && AbstractFurnaceBlockEntity.isFuel(inventory.get(FUEL_SLOT))) {
        litDuration = AbstractFurnaceBlockEntity.getFuel().get(inventory.get(FUEL_SLOT).getItem()) / 2;
        litTime = litDuration;
        inventory.get(FUEL_SLOT).shrink(1);
      }

      if(litTime > 0)
        processTime = Math.min(processTime + 1, PROCESS_TIME);
      else
        processTime = Math.max(processTime - 1, 0);

      if(processTime >= PROCESS_TIME) {
        var ingredient = inventory.get(INGREDIENT_SLOT);
        var data = ingredient.get(DataComponents.POTION_CONTENTS);
        if(data != null) {
          List<MobEffectInstance> effects = new ArrayList<>();
          for(var effect: data.getAllEffects()) {
            if(effect.isInfiniteDuration() || effect.getDuration() == 0) {
              effects.add(effect);
            } else {
              effects.add(new MobEffectInstance(
                effect.getEffect(),
                effect.getDuration() / 4,
                effect.getAmplifier() + 1
              ));
            }
          }
          var outputData = new PotionContents(data.potion(), data.customColor(), effects);
          var outputStack = new ItemStack(AlchimiaeItems.CONCENTRATE);
          outputStack.set(DataComponents.POTION_CONTENTS, outputData);

          if(inventory.get(RESULT_SLOT).isEmpty()) {
            inventory.set(RESULT_SLOT, outputStack);
            inventory.get(INGREDIENT_SLOT).shrink(1);
            processTime = 0;
          } else if(
            ItemStack.isSameItemSameComponents(outputStack, inventory.get(RESULT_SLOT)) &&
              inventory.get(RESULT_SLOT).getCount() < inventory.get(RESULT_SLOT).getMaxStackSize()
          ) {
            inventory.get(RESULT_SLOT).grow(1);
            inventory.get(INGREDIENT_SLOT).shrink(1);
            processTime = 0;
          }
        }
      }
    } else {
      processTime = Math.max(processTime - 3, 0);
    }
  }
}

File: common/src/main/kotlin/com/ssblur/alchimiae/blockentity/BoilerBlockEntity.kt
package com.ssblur.alchimiae.blockentity

import com.ssblur.alchimiae.data.AlchimiaeDataComponents
import com.ssblur.alchimiae.data.CustomEffect
import com.ssblur.alchimiae.data.CustomPotionEffects
import com.ssblur.alchimiae.item.AlchimiaeItems
import com.ssblur.alchimiae.network.client.AlchimiaeNetworkS2C
import com.ssblur.alchimiae.screen.menu.BoilerMenu
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.core.HolderLookup
import net.minecraft.core.NonNullList
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.ContainerHelper
import net.minecraft.world.WorldlyContainer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.entity.player.StackedContents
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ContainerData
import net.minecraft.world.inventory.StackedContentsCompatible
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.entity.AbstractFurnaceBlockEntity
import net.minecraft.world.level.block.entity.BaseContainerBlockEntity
import net.minecraft.world.level.block.entity.BlockEntity
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.Vec3
import java.util.*
import kotlin.math.max
import kotlin.math.min

class BoilerBlockEntity(blockPos: BlockPos, blockState: BlockState) :
  BaseContainerBlockEntity(AlchimiaeBlockEntities.BOILER.get(), blockPos, blockState),
  WorldlyContainer, StackedContentsCompatible {
  var litTime: Int = 0
  var litDuration: Int = 0
  var processTime: Int = 0
  val dataAccess: ContainerData

  var inventory: NonNullList<ItemStack>

  init {
    inventory = NonNullList.withSize(3, ItemStack.EMPTY)

    this.dataAccess = object : ContainerData {
      override fun get(i: Int): Int {
        return when (i) {
          0 -> {
            litTime
          }

          1 -> {
            litDuration
          }

          2 -> {
            processTime
          }

          else -> {
            0
          }
        }
      }

      override fun set(i: Int, j: Int) {
        when (i) {
          0 -> litTime = j
          1 -> litDuration = j
          2 -> processTime = j
        }
      }

      override fun getCount(): Int {
        return 3
      }
    }
  }

  override fun getContainerSize(): Int {
    return 3
  }

  override fun getDefaultName(): Component {
    return Component.translatable("menu.alchimiae.boiler")
  }

  override fun getItems(): NonNullList<ItemStack> {
    return inventory
  }

  override fun setItems(nonNullList: NonNullList<ItemStack>) {
    inventory = nonNullList
  }

  override fun isEmpty(): Boolean {
    return inventory.isEmpty()
  }

  override fun getItem(i: Int): ItemStack {
    return inventory[i]
  }

  override fun removeItem(i: Int, j: Int): ItemStack {
    setChanged()

    val stack = inventory[i]
    if (j >= stack.count) {
      inventory[i] = ItemStack.EMPTY
      return stack
    }
    stack.shrink(j)
    return stack.copyWithCount(j)
  }

  override fun removeItemNoUpdate(i: Int): ItemStack {
    val stack = inventory[i]
    inventory[i] = ItemStack.EMPTY
    return stack
  }

  override fun setItem(i: Int, itemStack: ItemStack) {
    inventory[i] = itemStack
  }

  override fun stillValid(player: Player): Boolean {
    return player.distanceToSqr(blockPos.center) < 16
  }

  override fun clearContent() {
    inventory = NonNullList.withSize(3, ItemStack.EMPTY)
  }

  override fun createMenu(i: Int, inventory: Inventory): AbstractContainerMenu {
    return BoilerMenu(i, inventory, this)
  }

  override fun getSlotsForFace(direction: Direction): IntArray {
    return when (direction) {
      Direction.DOWN -> {
        intArrayOf(RESULT_SLOT)
      }

      Direction.UP -> {
        intArrayOf(INGREDIENT_SLOT)
      }

      else -> {
        intArrayOf(FUEL_SLOT)
      }
    }
  }

  override fun canPlaceItemThroughFace(i: Int, itemStack: ItemStack, direction: Direction?): Boolean {
    return direction == null || Arrays.stream(getSlotsForFace(direction)).anyMatch { s: Int -> s == i }
  }

  override fun canTakeItemThroughFace(i: Int, itemStack: ItemStack, direction: Direction): Boolean {
    return Arrays.stream(getSlotsForFace(direction)).anyMatch { s: Int -> s == i }
  }

  override fun fillStackedContents(stackedContents: StackedContents) {
    for (itemStack in inventory) stackedContents.accountStack(itemStack)
  }

  override fun saveAdditional(compoundTag: CompoundTag, provider: HolderLookup.Provider) {
    super.saveAdditional(compoundTag, provider)
    ContainerHelper.saveAllItems(compoundTag, inventory, provider)

    compoundTag.putInt("processTime", processTime)
    compoundTag.putInt("litTime", litTime)
    compoundTag.putInt("litDuration", litDuration)
  }

  override fun loadAdditional(compoundTag: CompoundTag, provider: HolderLookup.Provider) {
    super.loadAdditional(compoundTag, provider)
    ContainerHelper.loadAllItems(compoundTag, inventory, provider)

    processTime = compoundTag.getInt("processTime")
    litTime = compoundTag.getInt("litTime")
    litDuration = compoundTag.getInt("litDuration")
  }

  fun tick() {
    if (level == null) return
    if (litTime > 0) {
      if (litTime % 3 == 0) {
        val x = blockPos.x + (10.0 / 16.0) + level!!.random.nextDouble() / 10.0
        val y = blockPos.y + (5.0 / 16.0)
        val z = blockPos.z + (5.0 / 16.0) + level!!.random.nextDouble() / 10.0
        if (level is ServerLevel)
          AlchimiaeNetworkS2C.flameParticle(Vec3(x, y, z), (level as ServerLevel).players())
      }
      litTime--
    }

    if (inventory[INGREDIENT_SLOT].`is`(AlchimiaeItems.MASH.get())) {
      if (litTime == 0 && AbstractFurnaceBlockEntity.isFuel(inventory[FUEL_SLOT])) {
        litDuration = AbstractFurnaceBlockEntity.getFuel()[inventory[FUEL_SLOT].item]!! / 2
        litTime = litDuration
        inventory[FUEL_SLOT].shrink(1)
      }

      processTime =
        if (litTime > 0) min(
          (processTime + 1).toDouble(),
          PROCESS_TIME.toDouble()
        )
          .toInt()
        else max((processTime - 1).toDouble(), 0.0).toInt()

      if (processTime >= PROCESS_TIME) {
        val ingredient = inventory[INGREDIENT_SLOT]
        val data = ingredient[AlchimiaeDataComponents.CUSTOM_POTION]
        if (data != null) {
          val effects: MutableList<CustomEffect> = ArrayList()
          for ((location, duration, strength) in data.effects) {
            if(duration <= 1) {
              effects.add(CustomEffect(location, duration, strength))
            } else {
              effects.add(CustomEffect(location, duration / 4, strength + 1))
            }
          }
          val outputStack = ItemStack(AlchimiaeItems.CONCENTRATE)
          outputStack[AlchimiaeDataComponents.CUSTOM_POTION] = CustomPotionEffects(effects, data.customColor)

          if (inventory[RESULT_SLOT].isEmpty) {
            inventory[RESULT_SLOT] = outputStack
            inventory[INGREDIENT_SLOT].shrink(1)
            processTime = 0
          } else if (ItemStack.isSameItemSameComponents(
              outputStack,
              inventory[RESULT_SLOT]
            ) &&
            inventory[RESULT_SLOT].count < inventory[RESULT_SLOT].maxStackSize
          ) {
            inventory[RESULT_SLOT].grow(1)
            inventory[INGREDIENT_SLOT].shrink(1)
            processTime = 0
          }
        }
      }
    } else {
      processTime = max((processTime - 3).toDouble(), 0.0).toInt()
    }
  }

  companion object {
    const val INGREDIENT_SLOT: Int = 0
    const val FUEL_SLOT: Int = 1
    const val RESULT_SLOT: Int = 2
    const val PROCESS_TIME: Int = 400

    @Suppress("unused")
    fun tick(level: Level?, blockPos: BlockPos?, blockState: BlockState?, blockEntity: BlockEntity) {
      if (blockEntity is BoilerBlockEntity) blockEntity.tick()
    }
  }
}
